scene/gui/tree.cpp

```diff
@@ -857,125 +812,80 @@ Tree *TreeItem::get_tree() const {
 }
 
 TreeItem *TreeItem::get_next() const {
-	return next;
+	if (parent == nullptr) {
+		return nullptr;
+	}
+	if (get_index() == parent->children.size() - 1) {
+		return nullptr;
+	}
+	return parent->children[
+		get_index() + 1
+	];
 }
 
 TreeItem *TreeItem::get_prev() {
-	if (prev) {
-		return prev;
-	}
-
-	if (!parent || parent->first_child == this) {
+	if (parent == nullptr) {
 		return nullptr;
 	}
-	// This is an edge case
-	TreeItem *l_prev = parent->first_child;
-	while (l_prev && l_prev->next != this) {
-		l_prev = l_prev->next;
+	if (get_index() == 0) {
+		return nullptr;
 	}
-
-	prev = l_prev;
-
-	return prev;
+	return parent->children[
+		get_index() - 1
+	];
 }
 
 TreeItem *TreeItem::get_parent() const {
 	return parent;
 }
 
 TreeItem *TreeItem::get_first_child() const {
-	return first_child;
+	if (children.size() == 0) {
+		return nullptr;
+	}
+	return children[0];
 }
 
 TreeItem *TreeItem::_get_prev_in_tree(bool p_wrap, bool p_include_invisible) {
-	TreeItem *current = this;
-
-	TreeItem *prev_item = current->get_prev();
-
-	if (!prev_item) {
-		current = current->parent;
-		if (current == tree->root && tree->hide_root) {
-			return nullptr;
-		} else if (!current) {
-			if (p_wrap) {
-				current = this;
-				TreeItem *temp = get_next_visible();
-				while (temp) {
-					current = temp;
-					temp = temp->get_next_visible();
-				}
-			} else {
-				return nullptr;
-			}
-		}
-	} else {
-		current = prev_item;
-		while ((!current->collapsed || p_include_invisible) && current->first_child) {
-			//go to the very end
-
-			current = current->first_child;
-			while (current->next) {
-				current = current->next;
+	if (get_index() != 0) {
+		for (uint64_t idx = get_index();idx > 0;idx--) {
+			TreeItem *current = parent->children[idx];
+			if (current->visible || p_include_invisible) {
```

Comment:
This doesn't look like the same algorithm at all to me, please explain how it does the same thing